playground/backend/cmd/server/controller.go

```diff
@@ -53,3 +111,183 @@ func (controller *playgroundController) GetCompileOutput(ctx context.Context, in
+	defer func(lc *fs_tool.LifeCycle) {
+		cancelByTimeoutFunc()
+		close(channel)
+		cleanUp(lc)
+	}(lc)
+
+	// validate
+	log.Println("parallelRunCode: Validate ...")
+	go exec.Validate(channel)
+	select {
+	case <-ctxWithTimeout.Done():
+		finishByContext(pipelineId)
+		return
+	case err := <-channel:
+		if err != nil {
+			// error during validation
+			processErrDuringValidate(err.(error), pipelineId)
+			return
+		}
+	}
+	log.Println("parallelRunCode: Validate finish")
+
+	// compile
+	log.Println("parallelRunCode: Compile ...")
+	go exec.Compile(channel)
+	select {
+	case <-ctxWithTimeout.Done():
+		finishByContext(pipelineId)
+		return
+	case err := <-channel:
+		if err != nil {
+			// error during compilation
+			processErrDuringCompile(err.(error), pipelineId)
+			return
+		}
+	}
+	log.Println("parallelRunCode: Compile finish")
+
+	// set empty value to pipelineId: cache.SubKey_CompileOutput
+	err := cacheService.SetValue(pipelineId, cache.SubKey_CompileOutput, "")
+	if err != nil {
+		log.Printf("parallelRunCode: cache.SetValue: " + err.Error())
+		return
+	}
+
+	// get executable file name
+	log.Println("parallelRunCode: get executable file name ...")
+	fileName, err := lc.GetExecutableName()
+	if err != nil {
+		// error during get executable file name
+		processErrDuringGetExecutableName(err, pipelineId)
+		return
+	}
+	log.Printf("parallelRunCode: executable file name: %s\n", fileName)
+
+	// run
+	output := ""
+	log.Println("parallelRunCode: Run ...")
+	go exec.Run(channel, fileName)
+	select {
+	case <-ctxWithTimeout.Done():
+		finishByContext(pipelineId)
+		return
+	case runResult := <-channel:
+		err := runResult.(*executors.RunResult).Err
+		output = runResult.(*executors.RunResult).Output
+		if err != nil {
+			// error during run code
+			processErrDuringRun(err, pipelineId)
+			return
+		}
+	}
+	log.Println("parallelRunCode: Run finish")
+	processSuccessRun(pipelineId, output)
+}
+
+// finishByContext is used in case of runCode method finished by timeout
+func finishByContext(pipelineId uuid.UUID) {
+	log.Println("parallelRunCode finish because of ctxWithTimeout.Done")
+
+	// set to cache pipelineId: cache.SubKey_Status: Status_STATUS_TIMEOUT
+	err := cacheService.SetValue(pipelineId, cache.SubKey_Status, pb.Status_STATUS_ERROR)
+	if err != nil {
+		log.Printf("parallelRunCode: cache.SetValue: " + err.Error())
+	}
+}
+
+// cleanUp removes all prepared folders for received LifeCycle
+func cleanUp(lc *fs_tool.LifeCycle) {
+	log.Println("parallelRunCode complete")
+
+	log.Println("parallelRunCode: DeleteFolders ...")
+	err := lc.DeleteFolders()
+	if err != nil {
+		log.Printf("parallelRunCode: DeleteFolders: " + err.Error())
+	}
+	log.Println("parallelRunCode: DeleteFolders complete")
+}
+
+// processErrDuringValidate processes error received during Validate step
+func processErrDuringValidate(err error, pipelineId uuid.UUID) {
+	log.Printf("parallelRunCode: Validate: " + err.Error())
```

Comment:
Let's add `pipelineId` for these logs too.

Reply:
Changed.